File: projects/m2/017-two-dice-simulation/solutions/EvaLateltin/017-two-dice-simulation.js
/*
LOGIC
Create an object that stores the result of the dices. At the beginning, the value for each key will be 0 (later it will be increased).
Create a function that simulate the roll of two dices; the function return their sum. 
-> create a random number from 1 to 6 for each dice
-> return the sum of each dice

In the object, add the value 1 to the key that equals the result of the rolls of dices.

Create a loop that: 
- calls 1000 times the function that rolls the dices
- each time, it returns the frequency of each total (absolute value and percentage value)
- also return the expected probability
*/

let rolls = 0;
const rollResults = {
  2: 0,
  3: 0,
  4: 0,
  5: 0,
  6: 0,
  7: 0,
  8: 0,
  9: 0,
  10: 0,
  11: 0,
  12: 0,
};

const rollProbability = {
  2: 1 / 36,
  3: 2 / 36,
  4: 3 / 36,
  5: 4 / 36,
  6: 5 / 36,
  7: 6 / 36,
  8: 5 / 36,
  9: 4 / 36,
  10: 3 / 36,
  11: 2 / 36,
  12: 1 / 36,
};

function rollOfDice() {
  return Math.floor(Math.random() * 6) + 1;
}

function sumOfTwoDices() {
  return rollOfDice() + rollOfDice();
}

while (rolls < 1000) {
  const result = sumOfTwoDices();
  rollResults[result] += 1;
  rolls++;
  console.log(
    `The sum of the dices is ${result}. 
In ${rolls} rolls, we got ${rollResults[result]} times the value ${result} - ${(
      (rollResults[result] / rolls) *
      100
    ).toFixed(2)}% of times. 
The expected probability to get ${result} is ${(
      rollProbability[result] * 100
    ).toFixed(2)}%.`
  );
}

File: projects/m2/017-two-dice-simulation/solutions/ericadegiovanni/017_two_dice_simulation.py
import random

def roll_2dice():

   # simulate the rolling of two dice (no parameters)
   # return the total that was rolled as its only result. 
    return random.randint(1,6) + random.randint(1,6)

def calc_occurrences(rolls):

    # return a dictionary with the times each key occurs in the given list

    results_counter = dict()
    for result in rolls:
        if result not in results_counter:
            results_counter[result] = 1
        else:
            results_counter[result] += 1
    return results_counter


def main():
    
    # count the times each possible result can occur from rolling two six-sided dice
    dice = [i for i in range(1,7)]
    all_possible_outcomes = [side1 + side2 for side1 in dice for side2 in dice ]
    outcomes_occurrence = calc_occurrences(all_possible_outcomes)

    # simulate 1000 times the roll of the dice
    roll_results = [roll_2dice() for _ in range(1, 1001)]      
    rolls_occurences = calc_occurrences(roll_results)

    # Create a table:
    # Express the frequency for each total as a percentage of the number of rolls performed. 
    # Display the percentage expected by probability  

    print('|         | Rolls     |  Probability |')
    print('| Result  | Frequency |  Theory      |')
    print('--------------------------------------')

    for i in range(2,13):
        print(f'    {i}'.ljust(12), end=' ')
        print(f'{(rolls_occurences[i]/len(roll_results))*100 :.2f}'.ljust(9), end=' ')
        print(f'   {(outcomes_occurrence[i]/len(all_possible_outcomes))*100 :.2f}')
    

if __name__ == '__main__':
    main()
